### ISSUE TITLE
Dom + HTTP: How to synchronize?

### ISSUE THREAD

**Issue by aslakhellesoy** [subsecondtdd/codebreaker-js#2]
The tests are currently failing because a step can still be “in flight” when the next one starts.

For example, the maker starting a game, and the breaker being unable to join it because it tries to join the game before the server has created it.

I think this might work:

* Every updateable “value” in the system has a version number
* The game list is a value
* Each game is a value
* There are 3 actors in each scenario: Maker, Breaker and the server
* The server is the only actor that performs value updates. (The Maker and Breaker don’t do updates, they just tell the server to update values).

Because there is only a single “updating” actor we don’t need to use version vectors. A simple version number should be sufficient.

Before a Maker creates a game, it stores the current version of the game list (which is 0).
After a createGame invocation, we wait for all actors to have a game list version equal to current version + 1.

The World knows about all the nodes, and can do this. We have to explicitly do this in every step.

How do we detect that a Maker/Breaker’s version number for a value has changed when we’re using DomSession?

I think we could use a DOM element that gets updated with the value’s version when the UI is rendered. We can detect this update with [DOM mutation events](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).

I’d love to hear what you think about this approach @tooky @joshski. I plan to hack on it tonight in case you’re available.

**Comment by joshski** [comment]
I definitely like the idea in general. I've always wanted to do something like this but never done it, but every time I see a browser test "eventually" fail it makes me sad.

> I think we could use a DOM element that gets updated with the value’s version when the UI is rendered. We can detect this update with DOM mutation events.

Do we really need DOM mutation events? If the test machinery is waiting for every client to have "rendered at a particular version of the server state" it can use polling. Polling is simpler and works in older browsers.

**Comment by aslakhellesoy** [comment]
The test machinery runs in Electron, so we don't have to worry about older browsers - it's always going to be whatever version of chrome Electron ships with. We're not explicitly firing the event from the app code, so nothing in the app needs to depend on the mutation event API. It's only the test code that listens for it.

Polling will also work, but it is slower, so why settle for next best? For Selenium we'd probably have to poll though, but that's a different issue.

**Comment by joshski** [comment]
> Polling will also work, but it is slower, so why settle for next best? For Selenium we'd probably have to poll though, but that's a different issue.

Yep fair enough, good point. I actually made a DOM implementation of "eventually", but based on mutation events in another project. I thought it would speed things up, and it did, a little bit, but not as much as I was expecting.
